tap_mailchimp/sync.py

```diff
@@ -1,541 +1,60 @@
-import json
-import time
-import random
-import tarfile
-
 import singer
-from singer import metrics, metadata, Transformer
-from singer.utils import strptime_to_utc, should_sync_field
-from requests.exceptions import HTTPError
+from tap_mailchimp.streams import STREAMS
 
 LOGGER = singer.get_logger()
 
-MIN_RETRY_INTERVAL = 2 # 2 seconds
-MAX_RETRY_INTERVAL = 300 # 5 minutes
-MAX_RETRY_ELAPSED_TIME = 43200 # 12 hours
-
-# Break up reports_email_activity batches to iterate over chunks
-EMAIL_ACTIVITY_BATCH_SIZE = 100
-
-class BatchExpiredError(Exception):
-    pass
-
-def next_sleep_interval(previous_sleep_interval):
-    min_interval = previous_sleep_interval or MIN_RETRY_INTERVAL
-    max_interval = previous_sleep_interval * 2 or MIN_RETRY_INTERVAL
-    return min(MAX_RETRY_INTERVAL, random.randint(min_interval, max_interval))
-
-def write_schema(catalog, stream_name):
-    stream = catalog.get_stream(stream_name)
-    schema = stream.schema.to_dict()
-    singer.write_schema(stream_name, schema, stream.key_properties)
-
-def process_records(catalog,
-                    stream_name,
-                    records,
-                    persist=True,
-                    bookmark_field=None,
-                    max_bookmark_field=None):
-    stream = catalog.get_stream(stream_name)
-    schema = stream.schema.to_dict()
-    stream_metadata = metadata.to_map(stream.metadata)
-    with metrics.record_counter(stream_name) as counter, Transformer() as transformer:
-        for record in records:
-            if bookmark_field:
-                if max_bookmark_field is None or \
-                   record[bookmark_field] > max_bookmark_field:
-                    max_bookmark_field = record[bookmark_field]
-            if persist:
-                record = transformer.transform(record,
-                                               schema,
-                                               stream_metadata)
-                singer.write_record(stream_name, record)
-                counter.increment()
-        return max_bookmark_field
-
-def get_bookmark(state, path, default):
-    dic = state
-    for key in ['bookmarks'] + path:
-        if key in dic:
-            dic = dic[key]
+def get_streams_to_sync(catalog, selected_streams, selected_stream_names):
+    """Return streams to sync"""
+    # List of top-level stream names to sync
+    streams_to_sync_names = []
+    # List of child stream names to sync if the grandchild is selected and the child is not selected
+    child_streams_to_sync = []
+    for stream in selected_streams:
+        # Get parent stream
+        parent_streams = STREAMS.get(stream.tap_stream_id).streams_to_sync
+        if parent_streams:
```

Comment:
Instead of it iterate through all the streams and check if the stream is selected or any child is selected or not.
This will lead to fewer condition checks.

Reply:
The `catalog.get_selected_streams` returns the list of stream in manner of currently_syncing ([Reference](https://github.com/singer-io/singer-python/blob/master/singer/catalog.py#L150)). Thus, we are looping on selected_streams rather than STREAMS.